pkg/spanconfig/spanconfigsqltranslator/sqltranslator.go

```diff
@@ -32,112 +33,129 @@ import (
-	execCfg *sql.ExecutorConfig
-	codec   keys.SQLCodec
-	knobs   *spanconfig.TestingKnobs
+	ptsProvider protectedts.Provider
+	codec       keys.SQLCodec
+	knobs       *spanconfig.TestingKnobs
+
+	txn      *kv.Txn
+	descsCol *descs.Collection
+	used     bool
+}
+
+// SQLTranslatorFactory is used to construct single-use, transaction-scoped
+// spanconfig.SQLTranslators.
+type SQLTranslatorFactory struct {
+	ptsProvider protectedts.Provider
+	codec       keys.SQLCodec
+	knobs       *spanconfig.TestingKnobs
 }
 
-// New constructs and returns a SQLTranslator.
-func New(
-	execCfg *sql.ExecutorConfig, codec keys.SQLCodec, knobs *spanconfig.TestingKnobs,
-) *SQLTranslator {
+// NewFactory constructs and returns a SQLTranslatorFactory.
+func NewFactory(
+	ptsProvider protectedts.Provider, codec keys.SQLCodec, knobs *spanconfig.TestingKnobs,
+) *SQLTranslatorFactory {
 	if knobs == nil {
 		knobs = &spanconfig.TestingKnobs{}
 	}
+	return &SQLTranslatorFactory{
+		ptsProvider: ptsProvider,
+		codec:       codec,
+		knobs:       knobs,
+	}
+}
+
+// SQLTranslator is part of the spanconfig.SQLTranslatorFactory interface.
+func (f *SQLTranslatorFactory) SQLTranslator(
+	txn *kv.Txn, descsCol *descs.Collection,
+) spanconfig.SQLTranslator {
 	return &SQLTranslator{
-		execCfg: execCfg,
-		codec:   codec,
-		knobs:   knobs,
+		ptsProvider: f.ptsProvider,
+		codec:       f.codec,
+		knobs:       f.knobs,
+		txn:         txn,
+		descsCol:    descsCol,
 	}
 }
 
 // Translate is part of the spanconfig.SQLTranslator interface.
 func (s *SQLTranslator) Translate(
 	ctx context.Context, ids descpb.IDs, generateSystemSpanConfigurations bool,
-) ([]spanconfig.Record, hlc.Timestamp, error) {
-	var records []spanconfig.Record
-	// txn used to translate the IDs, so that we can get its commit timestamp
-	// later.
-	var translateTxn *kv.Txn
-	if err := sql.DescsTxn(ctx, s.execCfg, func(
-		ctx context.Context, txn *kv.Txn, descsCol *descs.Collection,
-	) error {
-		// We're in a retryable closure, so clear any records from previous
-		// attempts.
-		records = records[:0]
-
-		// Construct an in-memory view of the system.protected_ts_records table to
-		// populate the protected timestamp field on the emitted span configs.
-		//
-		// TODO(adityamaru): This does a full table scan of the
-		// `system.protected_ts_records` table. While this is not assumed to be very
-		// expensive given the limited number of concurrent users of the protected
-		// timestamp subsystem, and the internal limits to limit the size of this
-		// table, there is scope for improvement in the future. One option could be
-		// a rangefeed-backed materialized view of the system table.
-		ptsState, err := s.execCfg.ProtectedTimestampProvider.GetState(ctx, txn)
+) (records []spanconfig.Record, _ hlc.Timestamp, _ error) {
+	if s.used {
+		return nil, hlc.Timestamp{}, errors.AssertionFailedf("SQLTranslator already used")
+	}
+
+	// Mark the Translator as used to disallow subsequent calls to Translate.
```

Comment:
Why make this single use? This variable's also making this unsafe for concurrent use. I can imagine wanting to paginate through a list of IDs at the caller now that it's all txn scoped -- making this single use gets in the way of that. And also makes it unsafe for concurrent only to serve an assertion.